apps/glusterfs/operations_manage.go

```diff
@@ -14,60 +14,96 @@ import (
 	"net/http"
 	"sync"
 
+	"github.com/lpabon/godbc"
+
 	"github.com/heketi/heketi/executors"
+	"github.com/heketi/heketi/pkg/idgen"
+)
 
-	"github.com/lpabon/godbc"
+type OpClass int
+
+const (
+	TrackNormal OpClass = iota
+	TrackToken
 )
 
-// OpCounter is used to track and manage how many operations are being
+// OpTracker is used to track and manage how many operations are being
 // processed by the server.
-type OpCounter struct {
+type OpTracker struct {
 	// configuration
 	Limit uint64
 
 	// internals
-	lock  sync.RWMutex
-	value uint64
+	lock      sync.RWMutex
+	normalOps map[string]bool
 }
 
-// Inc increments the internal operations counter.
-func (oc *OpCounter) Inc() {
-	oc.lock.Lock()
-	defer oc.lock.Unlock()
-	oc.value++
+func newOpTracker(limit uint64) *OpTracker {
+	return &OpTracker{
+		Limit:     limit,
+		normalOps: make(map[string]bool),
+	}
 }
 
-// Dec decrements the internal operations counter.
-func (oc *OpCounter) Dec() {
-	oc.lock.Lock()
-	defer oc.lock.Unlock()
-	godbc.Require(oc.value > 0)
-	oc.value--
+func (ot *OpTracker) insert(id string, c OpClass) {
+	godbc.Require(id != "", "id must not be empty")
+	godbc.Require(!ot.normalOps[id], "id already tracked", id)
+	ot.normalOps[id] = true
 }
 
-// Get returns the current value of the internal operations counter.
-func (oc *OpCounter) Get() uint64 {
-	oc.lock.RLock()
-	defer oc.lock.RUnlock()
-	return oc.value
+// Add records a new in-flight operation.
+func (ot *OpTracker) Add(id string, c OpClass) {
+	ot.lock.Lock()
+	defer ot.lock.Unlock()
+	ot.insert(id, c)
 }
 
-// ThrottleOrInc returns true if incrementing the counter would put
-// the number of operations over the limit, otherwise it increments
-// the counter and returns false. ThrottleOrInc exists to perform
+// Remove removes an operation from tracking.
+func (ot *OpTracker) Remove(id string) {
+	ot.lock.Lock()
+	defer ot.lock.Unlock()
+	godbc.Require(id != "", "id must not be empty")
+	godbc.Require(ot.normalOps[id], "id not tracked", id)
+	delete(ot.normalOps, id)
+}
+
+// Get returns the number of operations currently tracked.
+func (ot *OpTracker) Get() uint64 {
+	ot.lock.RLock()
+	defer ot.lock.RUnlock()
+	return uint64(len(ot.normalOps))
+}
+
+// ThrottleOrAdd returns true if adding the operation would put
+// the number of operations over the limit, otherwise it adds
+// the operation and returns false. ThrottleOrAdd exists to perform
 // the check and set atomically.
-func (oc *OpCounter) ThrottleOrInc() bool {
-	oc.lock.Lock()
-	defer oc.lock.Unlock()
-	if oc.value >= oc.Limit {
+func (ot *OpTracker) ThrottleOrAdd(id string, c OpClass) bool {
+	ot.lock.Lock()
+	defer ot.lock.Unlock()
+	n := len(ot.normalOps)
```

Comment:
cant we make use of `ot.Get()` here to avoid write lock in the whole function execution?

Reply:
No, the action of throttle or add must be atomic and so the check must be under the same lock as any changes we make.